src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java

```diff
@@ -0,0 +1,84 @@
+
+@Service
+@Transactional(readOnly = true)
+@Validated
+@Slf4j
+public class OffenderTransactionHistoryService {
+
+    private String apiCurrency;
+    private OffenderTransactionHistoryRepository repository;
+
+    public OffenderTransactionHistoryService(@Value("${api.currency:GBP}") final String currency,
+                                             final OffenderTransactionHistoryRepository repository) {
+        this.apiCurrency = currency;
+        this.repository = repository;
+    }
+
+    @VerifyOffenderAccess
+    public List<OffenderTransactionHistoryDto> getTransactionHistory(final Long offenderId,
+                                                                     final Optional<String> accountCodeOpl,
+                                                                     final Optional<LocalDate> fromDateOpl,
+                                                                     final Optional<LocalDate> toDateOpl) {
+
+        checkNotNull(offenderId, "offender-id can't be null");
+        checkNotNull(accountCodeOpl, "accountCode optional can't be null");
+        checkNotNull(fromDateOpl, "fromDate optional can't be null");
+        checkNotNull(toDateOpl, "toDate optional can't be null");
+
+        var fromDate = fromDateOpl.orElse(LocalDate.now());
+        var toDate = toDateOpl.orElse(LocalDate.now());
+
+        var now = LocalDate.now();
+        checkState(fromDate.isBefore(toDate) || fromDate.isEqual(toDate), "toDate can't be before fromDate");
+        checkState(fromDate.isBefore(now) || fromDate.isEqual(now), "fromDate can't be in the future");
+        checkState(toDate.isBefore(now) || toDate.isEqual(now), "toDate can't be in the future");
+
+        if(accountCodeOpl.isPresent()) {
+            boolean isAccountCodeExists = accountCodeOpl.map(AccountCode::byCodeName).filter(opl -> opl.isPresent()).isPresent();
+            checkState(isAccountCodeExists, "Unknown account-code " + accountCodeOpl.get());
+        }
+
+        var histories = (List<OffenderTransactionHistory>) accountCodeOpl
+                .map(AccountCode::byCodeName)
+                .filter(Optional::isPresent)
+                .map(optionalCode -> optionalCode.get().code)
+                .map(code -> repository.findForGivenAccountType(offenderId, code, fromDate, toDate))
+                .orElse(repository.findForAllAccountTypes(offenderId, fromDate, toDate));
+
+        var sortPolicy = Comparator
+                .comparing(OffenderTransactionHistory::getEntryDate)
+                .thenComparing(Comparator.comparing(OffenderTransactionHistory::getTransactionEntrySequence).reversed())
+                .reversed();
+
+        Collections.sort(histories, sortPolicy);
+
+        return histories.stream()
+                .map(h -> Pair.of(h, apiCurrency))
+                .map(OffenderTransactionHistoryTransformer::transform)
+                .collect(Collectors.toList());
+    }
+}
```

Comment:
This method is a bit long - could it be broken up?

Reply:
I've moved out the sort policy and it reduced it a bit but will need significant rework that makes sense to reduce it further

Reply:
It's just hard to see what is happening at a glance, so I'd be tempted to chop it up into at least 2 private methods (`validate` and `getSortedHistories`?).  But maybe that's just personal preference...

Reply:
done